src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java

```diff
@@ -0,0 +1,199 @@
+      }
+    }
+  }
+
+  /**
+   * Takes an Exception Object and relates its results to the user within the client.
+   *
+   * @param e The exception to be related.
+   * @throws Exception if there is some issue in the sending of the error messages.
+   */
+  protected void handleError(Exception e) throws Exception {
+    new ErrorResponse(this.outputStream, e, State.InternalError).send();
+    new ReadyResponse(this.outputStream, ReadyResponse.Status.IDLE).send();
+  }
+
+  /**
+   * Sends the result of an execute or query to the client. The type of message depends on the type
+   * of result of the statement. This method may also be called multiple times for one result if the
+   * client has set a max number of rows to fetch for each execute message. The {@link
+   * IntermediateStatement} will cache the result in between calls and continue serving rows from
+   * the position it was left off after the last execute message.
+   */
+  public boolean sendSpannerResult(IntermediateStatement statement,
+      QueryMode mode,
+      long maxRows)
+      throws Exception {
+    String command = statement.getCommand();
+    switch (statement.getResultType()) {
+      case NO_RESULT:
+        new CommandCompleteResponse(this.outputStream, command).send();
+        return false;
+      case RESULT_SET:
+        SendResultSetState state = sendResultSet(statement, mode, maxRows);
+        statement.setHasMoreData(state.hasMoreRows());
+        if (state.hasMoreRows()) {
+          new PortalSuspendedResponse(this.outputStream).send();
+        } else {
+          statement.getStatementResult().close();
+          new CommandCompleteResponse(
+              this.outputStream, "SELECT " + state.getNumberOfRowsSent()).send();
+        }
+        return state.hasMoreRows();
+      case UPDATE_COUNT:
+        command += ("INSERT ".equals(command)? "0 " : "") + statement.getUpdateCount();
+        new CommandCompleteResponse(this.outputStream, command).send();
+        return false;
+      default:
+        throw new IllegalStateException(
+            "Unknown result type: " + statement.getResultType());
+    }
+  }
+
+  /**
+   * Simple Adapter, which takes specific results from Spanner, and packages them in a format
+   * Postgres understands.
+   *
+   * @param describedResult Statement output by Spanner.
+   * @param mode Specific Query Mode required for this specific message for Postgres
+   * @param maxRows Maximum number of rows requested
+   *
+   * @return An adapted representation with specific metadata which PG wire requires.
+   * @throws Exception
+   */
+  public SendResultSetState sendResultSet(IntermediateStatement describedResult,
+      QueryMode mode,
+      long maxRows) throws Exception {
+    long rows = 0;
+    boolean hasData = describedResult.isHasMoreData();
+    ResultSet resultSet = describedResult.getStatementResult();
+    while (hasData) {
+      new DataRowResponse(
```

Comment:
ditto before - this looks like the tight loop for sending a result set.  Might need to be optimized.  Ok to do in a subsequent PR.

Reply:
Added todo